uniffi_core/src/ffi/foreigncallbacks.rs

```diff
@@ -195,9 +197,160 @@ impl ForeignCallbackInternals {
         };
     }
 
-    pub fn get_callback(&self) -> Option<ForeignCallback> {
+    fn call_callback(
+        &self,
+        handle: u64,
+        method: u32,
+        args: RustBuffer,
+        ret_rbuf: &mut RustBuffer,
+    ) -> c_int {
         let ptr_value = self.callback_ptr.load(Ordering::SeqCst);
-        unsafe { std::mem::transmute::<usize, Option<ForeignCallback>>(ptr_value) }
+        unsafe {
+            let callback = std::mem::transmute::<usize, Option<ForeignCallback>>(ptr_value)
+                .expect("Callback interface handler not set");
+            callback(
+                handle,
+                method,
+                args.data_pointer(),
+                args.len() as i32,
+                ret_rbuf,
+            )
+        }
+    }
+
+    /// Invoke a callback interface method that can't throw on the foreign side and returns
+    /// non-Result value on the Rust side
+    pub fn invoke_callback<R, UniFfiTag>(&self, handle: u64, method: u32, args: RustBuffer) -> R
+    where
+        R: FfiConverter<UniFfiTag>,
+    {
+        let mut ret_rbuf = RustBuffer::new();
+        let callback_result = self.call_callback(handle, method, args, &mut ret_rbuf);
+        match callback_result {
+            1 => {
+                // 1 indicates success with the return value written to the RustBuffer for
+                //   non-void calls.
+                let vec = ret_rbuf.destroy_into_vec();
+                let mut ret_buf = vec.as_slice();
+                R::try_read(&mut ret_buf).unwrap()
+            }
+            -2 => {
+                panic!("Callback return -2, but no Error was expected")
+            }
+            // 0 is a deprecated method to indicates success for void returns
+            0 => {
```

Comment:
That makes a ton of sense.  In fact, since we're breaking the ABI,  I think we should change the codes to make a bit more sense (success = 0, error = 1, unexpected error = 2).